### PR TITLE
Using pyupgrade to convert string formatting to `f-string`.

### LINKED ISSUE #3496: Use `f-strings` everywhere for string formatting/concatenation
I've noticed that PyBaMM uses different (`+`, `.format`, and `f-strings`) string operations in different places. We should replace all the `+` and `.format` operations with `f-strings`. `f-strings` are more efficient (than `+`) and much more readable (than `+` and `.format`).

[[PEP 498](https://peps.python.org/pep-0498/)]

### PR DESCRIPTION
**Pull request by prady0t**
Fixes #3496 

Linked to PR #3551

### CODE REVIEW

**Review by Saransh-cpp — CHANGES_REQUESTED** [review]
Thanks for working on this, @prady0t!

1. Not all pre-commit related changes were included in https://github.com/pybamm-team/PyBaMM/pull/3579/commits/41019cf34259c2c218f48d847eef75d261759c37, so you might have to reset the commits and include all the changes in a single commit.
2. Some pre-commit suggestions must be resolved manually. These are visible here - https://results.pre-commit.ci/run/github/155538761/1701378735.92sVUq40SsSDBx8YpLEX4w. You can access this by scrolling to the bottom of your PR and clicking "details" on the pre-commit.ci job.
3. There are a few typing fixes required for the pre-commit job to pass. Adding `UP007` to the list of excluded checks in `pyproject.toml` with a small note should fix this for now. (@pipliggins could fix these and add the check back in #3578.)

**Comment by prady0t** [comment]
> Thanks for working on this, @prady0t!
> 
> 1. Not all pre-commit related changes were included in [41019cf](https://github.com/pybamm-team/PyBaMM/commit/41019cf34259c2c218f48d847eef75d261759c37), so you might have to reset the commits and include all the changes in a single commit.
> 2. Some pre-commit suggestions must be resolved manually. These are visible here - https://results.pre-commit.ci/run/github/155538761/1701378735.92sVUq40SsSDBx8YpLEX4w. You can access this by scrolling to the bottom of your PR and clicking "details" on the pre-commit.ci job.
> 3. There are a few typing fixes required for the pre-commit job to pass. Adding `UP007` to the list of excluded checks in `pyproject.toml` with a small note should fix this for now. (@pipliggins could fix these and add the check back in [Adds typing to expression tree #3578](https://github.com/pybamm-team/PyBaMM/pull/3578).)

Am I supposed to add `UP007` here? https://github.com/pybamm-team/PyBaMM/blob/ebacf496a4322b084b7c4058599755b45008283b/pyproject.toml#L202 also how do I add commit hash (mentioned earlier)?

**Comment by kratman** [comment]
@prady0t How is this coming along?

**Comment by prady0t** [comment]
> @prady0t How is this coming along?

Sorry for not being active. I had exams then had to travel back home. Was working on this just today, will complete this by EOD.

**Comment by kratman** [comment]
No rush, just curious since you were close

**Comment by prady0t** [comment]
Done! Please have a look.

**Comment by agriyakhetarpal** [comment]
After you reset your branch locally you pulled the changes in it that were present on GitHub instead of a force push (I think). This is why 101 or so commits are showing up (probably due to a broken merge commit, the presence of which would have fixed it otherwise)

**Comment by prady0t** [comment]
I might have to close this PR as well. What do you think?

**Comment by agriyakhetarpal** [comment]
You can reset your branch and keep working on the same PR using the instructions posted above, no worries! Otherwise, you can undo your commits till https://github.com/pybamm-team/PyBaMM/pull/3579/commits/75b58bc50646e5599027e73f5c8254714a20754e is the last commit.

**Comment by kratman** [comment]
@prady0t You are so close. If you want I can post the full set of commands to do it with reset/cherry pick. The general procedure I would follow is
- Reset to origin/main
- Cherry pick your pre-commit config changes from this branch on your fork's origin (origin/new-format-fix)
- Run pre-commit, commit formatting changes, copy the commit hash
- Add the formatting changes hash to the ignored revs, and commit that change
- Force push to this branch

**Comment by Saransh-cpp** [comment]
You can also `git log` the hashes of your last 4 commits, reset everything in this branch, and then `git cherry-pick` those 4 commits using their hashes.

This would be a nice git exercise but please feel free to open a new PR if you feel that would be better.

**Comment by prady0t** [comment]
Did this work? I basically just did this:
```
git reset --hard 75b58bc  
git push origin -f
```

**Comment by agriyakhetarpal** [comment]
Yes, you are back on track. All that is left is to follow the instructions in https://github.com/pybamm-team/PyBaMM/pull/3579#issuecomment-1847220060

**Comment by prady0t** [comment]
> Yes, you are back on track. All that is left is to follow the instructions in [#3579 (comment)](https://github.com/pybamm-team/PyBaMM/pull/3579#issuecomment-1847220060)

But all those commits are already there and ci test have passed as well. Do I have to do it again?

File: pybamm/solvers/processed_variable.py [hunk 1]
```diff
@@ -106,7 +106,7 @@ def __init__(
                     else:
                         # Raise error for 3D variable
                         raise NotImplementedError(
-                            "Shape not recognized for {} ".format(base_variables[0])
+                            f"Shape not recognized for {base_variables[0]} "
```
Comment (agriyakhetarpal) [inline pybamm/solvers/processed_variable.py:109]:
```suggestion
                            f"Shape not recognized for {base_variables[0]}"
```

**Review by agriyakhetarpal — APPROVED** [review]
We should be done, then – these changes look great! Thanks, @prady0t!

File: pyproject.toml [hunk 2]
```diff
@@ -196,7 +196,7 @@ extend-select = [
   "RUF",         # Ruff-specific
   # "SIM",         # flake8-simplify
   # "T20",         # flake8-print
-  # "UP",          # pyupgrade
+   "UP",          # pyupgrade
   "YTT",         # flake8-2020
```
Comment (Saransh-cpp) [inline pyproject.toml:200]:
```suggestion
  "UP",           # pyupgrade
  "YTT",         # flake8-2020
```

File: pyproject.toml [hunk 3]
```diff
@@ -214,6 +214,7 @@ ignore = [
   "RET506",      # Unnecessary `elif`
   "B018",        # Found useless expression
   "RUF002",      # Docstring contains ambiguous
+    "UP007",     # For pyupgrade
```
Comment (Saransh-cpp) [inline pyproject.toml:217]:
```suggestion
  "RUF002",      # Docstring contains ambiguous
  "UP007",       # For pyupgrade
```

File: pybamm/discretisations/discretisation.py [hunk 4]
```diff
@@ -19,7 +19,7 @@ def has_bc_of_form(symbol, side, bcs, form):
         return False
 
 
-class Discretisation(object):
```
Comment (Saransh-cpp) [inline pybamm/discretisations/discretisation.py:22]: Nice, getting rid of Python 2!

File: pybamm/discretisations/discretisation.py [hunk 5]
```diff
@@ -74,7 +72,7 @@ def y_slices(self):
     @y_slices.setter
     def y_slices(self, value):
         if not isinstance(value, dict):
-            raise TypeError("""y_slices should be dict, not {}""".format(type(value)))
+            raise TypeError(f"""y_slices should be dict, not {type(value)}""")
```
Comment (Saransh-cpp) [inline pybamm/discretisations/discretisation.py:75]:
```suggestion
            raise TypeError(f"y_slices should be dict, not {type(value)}")
```

File: pybamm/models/full_battery_models/base_battery_model.py [hunk 6]
```diff
@@ -1021,10 +1021,8 @@ def options(self, extra_options):
             and options["hydrolysis"] == "true"
         ):
             raise pybamm.OptionError(
-                """must use surface formulation to solve {!s} with hydrolysis
-                    """.format(
-                    self
-                )
+                f"""must use surface formulation to solve {self!s} with hydrolysis
+                    """
```
Comment (Saransh-cpp) [inline pybamm/models/full_battery_models/base_battery_model.py:1025]:
```suggestion
                f"must use surface formulation to solve {self!s} with hydrolysis"
```

**Review by Saransh-cpp — APPROVED** [review]
Looks good, thanks @prady0t! I could see some formatting issues but hopefully they'll be resolved once `ruff format` is added in.

**Comment by prady0t** [comment]
Thanks for being so supportive everyone 😄 .

**Review by kratman — APPROVED** [review]

**Comment by kratman** [comment]
https://github.com/all-contributors please add @prady0t for infrastructure

**Comment by kratman** [comment]
@all-contributors please add @prady0t for infrastructure

File: pyproject.toml [hunk 7]
```diff
@@ -196,7 +196,7 @@ extend-select = [
   "RUF",         # Ruff-specific
   # "SIM",         # flake8-simplify
   # "T20",         # flake8-print
-  # "UP",          # pyupgrade
+  "UP",           # pyupgrade
   "YTT",         # flake8-2020
```
Comment (Saransh-cpp) [inline pyproject.toml:200]: nitpicking 🙂 
```suggestion
  "UP",          # pyupgrade
  "YTT",         # flake8-2020
```

**Pull request merged**